waku/v2/protocol/waku_enr/sharding.nim

```diff
@@ -0,0 +1,222 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/[options, bitops, sequtils],
+  stew/[endians2, results],
+  stew/shims/net,
+  eth/keys,
+  libp2p/[multiaddress, multicodec],
+  libp2p/crypto/crypto
+import
+  ../../../common/enr,
+  ../waku_message
+
+
+const MaxShardIndex: uint16 = 1023
+
+const
+  ShardingIndicesListEnrField* = "rs"
+  ShardingBitVectorEnrField* = "rsv"
+
+
+type
+  RelayShards* = object
+    cluster: uint16
+    indices: seq[uint16]
+
+
+func cluster*(rs: RelayShards): uint16 =
+  rs.cluster
+
+func indices*(rs: RelayShards): seq[uint16] =
+  rs.indices
+
+func topics*(rs: RelayShards): seq[NsPubsubTopic] =
+  rs.indices.mapIt(NsPubsubTopic.staticSharding(rs.cluster, it))
+
+
+func init*(T: type RelayShards, cluster, index: uint16): T =
+  if index > MaxShardIndex:
+    raise newException(Defect, "invalid index")
+
+  RelayShards(cluster: cluster, indices: @[index])
+
+func init*(T: type RelayShards, cluster: uint16, indices: varargs[uint16]): T =
+  if toSeq(indices).anyIt(it > MaxShardIndex):
+    raise newException(Defect, "invalid index")
+
+  let indicesSeq = deduplicate(@indices)
+  if indices.len < 1:
+    raise newException(Defect, "invalid index count")
+
+  RelayShards(cluster: cluster, indices: indicesSeq)
+
+func init*(T: type RelayShards, cluster: uint16, indices: seq[uint16]): T =
+  if indices.anyIt(it > MaxShardIndex):
+    raise newException(Defect, "invalid index")
+
+  let indicesSeq = deduplicate(indices)
+  if indices.len < 1:
+    raise newException(Defect, "invalid index count")
+
+  RelayShards(cluster: cluster, indices: indicesSeq)
+
+
+func contains*(rs: RelayShards, cluster, index: uint16): bool =
+  rs.cluster == cluster and rs.indices.contains(index)
+
+func contains*(rs: RelayShards, topic: NsPubsubTopic): bool =
+  if topic.kind != NsPubsubTopicKind.StaticSharding:
+    return false
+
+  rs.contains(topic.cluster, topic.shard)
+
+func contains*(rs: RelayShards, topic: PubsubTopic|string): bool =
+  let parseRes = NsPubsubTopic.parse(topic)
+  if parseRes.isErr():
+    return false
+
+  rs.contains(parseRes.value)
+
+
+# ENR builder extension
+
+func toIndicesList(rs: RelayShards): seq[byte] =
+  if rs.indices.len > 255:
+    raise newException(Defect, "indices list too long")
```

Comment:
In this case I'd rather return a `Result[seq[byte], string]` and give as much detail as possible in case of error.

Reply:
Agree, changing it 👍🏼